master/internal/hpimportance/hpimportance.go

```diff
@@ -1,15 +1,227 @@
+		}
+
+		if record[1] == "metric" || record[1] == "numBatches" {
+			continue
+		}
+		hpi[record[1]], err = strconv.ParseFloat(record[2], 64)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse HP importance value: %s", err.Error())
+		}
+	}
+	return hpi, nil
+}
+
+// For the implementation, since we need to account for adaptive search
+// but we don't want to compare trials that have trained for 10,000 vs 100 batches
+// therefore, we continue to add trials till one of 2 conditions are met.
+// 1. There are at least 50 trials to train with and the ideal number of trial difference is met.
+// 2. The difference between the most(aka max) trained trial and the lowest batch are within a
+// defined range (maxDiffCompBatches).
+func computeHPImportance(data map[int][]model.HPImportanceTrialData,
+	experimentConfig *model.ExperimentConfig, masterConfig HPImportanceConfig,
+	growforest string, workingDir string) (map[string]float64, error) {
+	totalNumTrials, err := createDataFile(data, experimentConfig, workingDir)
+	if err != nil {
+		return nil, fmt.Errorf("error writing ARFF file: %s", err.Error())
+	}
+
+	nCores := strconv.FormatInt(int64(masterConfig.CoresPerWorker), 10)
+	maxNumTrees := int64(masterConfig.MaxTrees)
+
+	// random may be smaller because only 50 trials are ran
+	// where I'm not gonna calculate the random forest
+	// TODO: Determine best way to handle small amount of trials
+	if totalNumTrials < minNumberTrials {
+		return nil, fmt.Errorf("not enough trials for HP importance: " + fmt.Sprint(totalNumTrials))
+	}
+
+	// For version one, we do half the number of trials up to 300
+	// This may be overdoing and running too long for minimal performance inc.
+	// The paper uses 2 values replications and ntrees while the library,
+	// only uses --ace which is replications.
+	// TODO: use the bar chart from hp viz to improve hp importance
+	numTrees := int64(totalNumTrials / nReplications)
+
+	if numTrees > maxNumTrees {
+		numTrees = maxNumTrees
+	}
+	strNumTrees := strconv.FormatInt(numTrees, 10)
+
+	// Call Random Forest
+	// Ignore security warning because none of this is user-provided input
+	// #nosec G204
+	output, err := exec.Command(growforest,
+		// the data file created above
+		"-train", path.Join(workingDir, arffFile),
+		// the y value used to predict. We use a generic metric so we don't have to keep track
+		// of current metric since name doesn't matter
+		"-target", "metric",
+		// number of CPUs to use
+		"-nCores", nCores,
+		// number of replications/permutations
+		"-ace", strNumTrees,
+		// output file
+		"-importance", path.Join(workingDir, importanceFile),
+		// file name to output predictor forest in sf format.
+		"-rfpred", path.Join(workingDir, "rface.sf"),
+	).CombinedOutput()
+	if err != nil {
+		fmt.Printf("growforest failed:\n" + string(output))
+		return nil, fmt.Errorf("random forest failed: %s", err.Error())
+	}
+
+	hpi, err := parseImportanceOutput(path.Join(workingDir, importanceFile))
```

Comment:
same comment for this `path.Join`. construct it above and use it in both places maybe.